datatree/mapping.py

```diff
@@ -251,6 +256,23 @@ def _map_over_subtree(*args, **kwargs) -> DataTree | Tuple[DataTree, ...]:
     return _map_over_subtree
 
 
+def _handle_errors_with_path_context(path):
+    """Wraps given function so that if it fails it also raises path to node on which it failed."""
+
+    def decorator(func):
+        def wrapper(*args, **kwargs):
+            try:
+                return func(*args, **kwargs)
+            except Exception as e:
+                # Add the context information to the error message
+                e.add_note(f"Raised whilst mapping function over node with path {path}")
```

Comment:
Adding notes to exceptions is a new feature in python 3.11.

Perhaps we should wait before merging this? If I understand xarray's deprecation policy correctly then 3.11 will only be the minimum supported version of python 24 months after 3.11 is released, which is 12 months from now...

Reply:
you could use [a helper](https://github.com/agronholm/exceptiongroup/issues/31#issue-1404053238) for this:
```python
def add_note(err: BaseException, msg: str) -> None:
    # TODO: remove once python 3.10 can be dropped
    if sys.version_info < (3, 11):
        err.__notes__ = getattr(err, "__notes__", []) + [msg]
    else:
        err.add_note(msg)
```